feat: support api key for the api me template

huimiu opened the pull request:
https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/25643763

huimiu commented on templates/js/copilot-plugin-from-scratch-api-key/env/.env.dev:
```diff
@@ -0,0 +1,18 @@
+TEAMS_APP_TENANT_ID=
+API_FUNCTION_ENDPOINT=
+API_FUNCTION_RESOURCE_ID=
+DOTNET_PATH=
```
Thanks, the DOTNET_PATH variable has been removed.

huimiu commented on templates/js/copilot-plugin-from-scratch-api-key/env/.env.local:
```diff
@@ -0,0 +1,16 @@
+TEAMS_APP_ID=
+TEAMS_APP_PACKAGE_PATH=
+FUNC_ENDPOINT=
+DOTNET_PATH=
```
Removed.

huimiu commented on templates/js/copilot-plugin-from-scratch-api-key/keyGen.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+# The file path where the API Key is located.
+file_path="env/.env.$1.user"
```
Yes, we specify using Bash to handle the script in the script action, so it can run across different platforms.

```
  # Set required variables for local launch
  - uses: script
    with:
      shell: bash
      run:
        ./keyGen.sh ${{TEAMSFX_ENV}};
```

huimiu commented on templates/js/copilot-plugin-from-scratch-api-key/env/.env.local:
```diff
@@ -0,0 +1,16 @@
+TEAMS_APP_PACKAGE_PATH=
+FUNC_ENDPOINT=
+DOTNET_PATH=
+TEAMS_APP_TENANT_ID=
```
The additional API_FUNCTION_ENDPOINT and API_FUNCTION_RESOURCE_ID variables are the output of the Bicep file deployment, containing information related to the endpoint and resource ID after deploying Azure Function.

For TEAMS_APP_UPDATE_TIME, this should be placed in the .env.user file. I confirmed with the owner of this section, it will be automatically generated after provisioning and it will only be consumed internally by TTK. So, I have removed this variable.

MuyangAmigo commented on templates/js/copilot-plugin-from-scratch-api-key/README.md:
```diff
@@ -0,0 +1,59 @@
+# Overview of Custom Search Results app template
+
+## Build a message extension from a new API with Azure Functions
+
```
"Build a message extension from a new API with Azure Function and API key authentication"

huimiu replied:
Updated

MuyangAmigo reviewed (dismissed):
Approve for README

eriolchan commented on templates/js/copilot-plugin-from-scratch-api-key/teamsapp.yml.tpl:
```diff
@@ -0,0 +1,171 @@
+
+  - uses: cli/runNpmCommand
+    name: build app
+    env:
```
why run build need this env variable?

huimiu replied:
Removed this action from js template.

huimiu commented on templates/ts/copilot-plugin-from-scratch-api-key/teamsapp.local.yml.tpl:
```diff
@@ -0,0 +1,104 @@
+  # Set required variables for local launch
+  - uses: script
+    with:
+      shell: bash
```
Thanks for pointing out this issue. After testing, I found that running this action on Windows without WSL installed would result in an error. I've replaced the shell script with a JavaScript/TypeScript script, and running it through Node to achieve cross-platform capability.

huimiu commented:
After recent discussions, I made the following changes. Please help review again, thanks.

1. Replaced the script for generating API keys with a JavaScript/TypeScript script, using Node to achieve cross-platform functionality.
2. No longer automatically generating API keys for users. They will need to refer to the readme to manually generate them.
3. Keep the logic for validating API keys in the code unchanged. Added some comments in the code to explain its purpose and informed users on how to use Azure Functions' built-in authentication. We plan to release this template first and gather feedback to decide whether to modify it to leverage the Azure Functions' built-in authentication capabilities.

eriolchan commented on templates/js/copilot-plugin-from-scratch-api-key/package.json.tpl:
```diff
@@ -0,0 +1,17 @@
+        "dev": "func start --javascript --language-worker=\"--inspect=9229\" --port \"7071\" --cors \"*\"",
+        "start": "npx func start",
+        "test": "echo \"Error: no test specified\" && exit 1",
+        "key:gen": "node ./keygen.js",
```
Is there any reason to add :, why not keygen for task name?

huimiu replied:
Updated
